ideas/282-deterministic-builds.md

```diff
@@ -0,0 +1,53 @@
+---
+id: 282-deterministic-builds
+title: Deterministic Mobile App Builds
+status: In Progress
+created: 2018-08-16
+category: infra
+lead-contributor: jakubgs
+contributors:
+exit-criteria: yes
+success-metrics: yes
+clear-roles: yes
+future-iterations: yes
+roles-needed:
+---
+
+## Preamble
+
+    Idea: #282-deterministic-builds
+    Title: Deterministic Mobile App Builds
+    Status: In Progress
+    Created: 2018-08-16
+
+The purpose of achieving deterministic builds is encuring security of Status software. Deterministic builds depend on ensuring stable inputs, ensuring stable outputs, and capturing as little of the build environment as possible.
+
+In simple terms, deterministic builds mean the same commit generates exactly the same software package(s) regardless of where and when it is built. This is achieved by controlling all possible variables in a build, which includes controlling your build environment as well.
+
+## Objectives
+
+- status-go builds are deterministic
+- status-react builds are deterministic
+
+## Key Results
+
+- Anyone can build a commit on their own machines and verify the result by comparing against our builds
+- We can upload Android builds to F-Droid which requires deterministic builds
+
+## Timeline / Checkpoints
+
+__TODO__: Not sure?
+
+# Exit Criteria
+
+- All Status mobile and desktop app builds are deterministic
+
+## What exactly we going to do?
+
+- Verify all of our dependencies are frozen and versioned (`Gemfile.lock`)
+- Verify we depend on no resources pulled from internet during build
```

Comment:
Is it about resources from Internet which we do not control or about any resources? For example, desktop builds `status-go` by doing `git clone` on-the-go for specific revision. Usually, there is no issue with doing that.

Reply:
@jakubgs, during desktop mac build we also download a prepopulated zip file with libraries and frameworks needed to create `StatusIm.dmg`. Currently, it is stored in a separate github repository. Where should it be stored for a deterministic build?

Reply:
I believe that we should be able to automate creation process of prepopulated zip file content. The main concern is `ubuntu_server` binary with modified `realm` node-js module.

Reply:
@vkjr is right, this is the kind of external resources I'm talking about. Now as @MaxRis said, this should be either:

a) built by us in a deterministic way
b) versioned and checksumed so we know we get the same thing every time